src/python/pants/backend/codegen/tasks/simple_codegen_task.py

```diff
@@ -172,3 +217,119 @@ def execute(self):
         self.update_artifact_cache(vts_artifactfiles_pairs)
+
+
+  class CodegenStrategy(object):
+    """Abstract strategies for running codegen.
+
+    Includes predicting generated sources, partitioning targets for execution, etc.
+    """
+    __metaclass__ = ABCMeta
+
+    @property
+    def name(self):
+      """The name of this strategy (eg, 'isolated')."""
+      raise NotImplementedError
+
+    @abstractmethod
+    def execute_codegen(self, targets):
+      """Invokes the task's execute_codegen on the targets.
+
+      Subclasses decide how the targets are partitioned before being sent to the task's
+      execute_codegen method.
+      :targets: a set of targets.
+      """
+      raise NotImplementedError
+
+    @abstractmethod
+    def find_sources(self, target):
+      """Finds (or predicts) the sources generated by the given target."""
+      raise NotImplementedError
+
+    @abstractmethod
+    def codegen_workdir_suffix(self, target):
+      """The working directory suffix for the given target's generated code."""
+      raise NotImplementedError
+
+    def __str__(self):
+      return self.name
+
+
+  class GlobalCodegenStrategy(CodegenStrategy):
+    """Code generation strategy which generates all code together, in base directory."""
+
+    def __init__(self, task):
+      self._task = task
+
+    @property
+    def name(self):
+      return 'global'
+
+    def execute_codegen(self, targets):
+      self._task.execute_codegen(targets)
+
+    def find_sources(self, target):
+      return self._task.sources_generated_by_target(target)
+
+    def codegen_workdir_suffix(self, target):
+      return self.name
+
+
+  class IsolatedCodegenStrategy(CodegenStrategy):
+    """Code generate strategy which generates the code for each target separately.
+
+    Code is generated in a unique parent directory per target.
+    """
+
+    def __init__(self, task):
+      self._task = task
+      # NOTE(gm): This memoization yields a ~10% performance increase on my machine.
+      self._generated_sources_cache = {}
+
+    @property
+    def name(self):
+      return 'isolated'
+
+    def execute_codegen(self, targets):
+      for target in targets:
+        self._task.execute_codegen([target])
+
+    def find_sources(self, target):
+      return self._find_sources_strictly_generated_by_target(target)
+
+    def codegen_workdir_suffix(self, target):
+      return os.path.join(self.name, target.id)
+
+    def _find_sources_generated_by_target(self, target):
+      if target.id in self._generated_sources_cache:
+        for source in self._generated_sources_cache[target.id]:
+          yield source
+      target_workdir = self._task.codegen_workdir(target)
+      if not os.path.exists(target_workdir):
+        return
+      for root, dirs, files in safe_walk(target_workdir):
+        for name in files:
+          yield os.path.join(root, name)
+
+    def _find_sources_generated_by_dependencies(self, target, relative=False):
```

Comment:
Why the relative=False argument? I see only one caller where relative is forced to True.